### PR TITLE
Added retry 'registry' chain element

### LINKED ISSUE #1201: Implement 'retry' registry chain element
## Overview

To avoid problems with the concurrent start of nsmgr, forwarder, nse we need to implement 'retry' chain element for the registry. 

## Implementation details

1. Implement registry chain element 'retry'. See at example in https://github.com/networkservicemesh/sdk/tree/main/pkg/networkservice/common/retry
2. Add unit tests 
3. Add the chain element into https://github.com/networkservicemesh/sdk/tree/main/pkg/registry/chains/client
4. Check that everything is working

## References

Related to https://github.com/networkservicemesh/sdk/issues/1170
Related to https://github.com/networkservicemesh/cmd-forwarder-vpp/issues/435

### PR DESCRIPTION
**Pull request by Sol-0**
Signed-off-by: Oleg Solodkov <oleg.solodkov@xored.com>

## Issue link
Fixes #1201

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [x] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/registry/common/retry/options.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import "time"
+
+type configurable interface {
+	setInterval(time.Duration)
+	setTryTimeout(time.Duration)
+}
+
+// Option is NS/NSE retry client config option
+type Option interface {
+	apply(configurable)
```
Comment (denis-tingaikin) [inline pkg/registry/common/retry/options.go:28]:
```suggestion
	apply(*options)
```

File: pkg/registry/common/retry/options.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+
+package retry
+
+import "time"
+
+type configurable interface {
+	setInterval(time.Duration)
+	setTryTimeout(time.Duration)
+}
```
Comment (denis-tingaikin) [inline pkg/registry/common/retry/options.go:24]:
```suggestion
type options interface {
	internal time.Duration
	timeout time.Duration
}
```

File: pkg/registry/common/retry/options.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// Option is NS/NSE retry client config option
+type Option interface {
+	apply(configurable)
+}
+
+type applier func(configurable)
+
+func (f applier) apply(c configurable) {
+	f(c)
+}
```
Comment (denis-tingaikin) [inline pkg/registry/common/retry/options.go:35]:
```suggestion
```

File: pkg/registry/utils/count/nse_client.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+)
+
+// NSEClient is a client type for counting Register / Unregister / Find
+type NSEClient struct {
+	totalRegisterCalls, totalUnregisterCalls, totalFindCalls int32
+	registers, unregisters, finds                            map[string]int32
+	mu                                                       sync.Mutex
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/nse_client.go:36]: These fields are not used. Please don't add dead code.

```suggestion
```

File: pkg/registry/utils/count/nse_client.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+)
+
+// NSEClient is a client type for counting Register / Unregister / Find
+type NSEClient struct {
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/nse_client.go:33]:
```suggestion
type countNSEClient struct {
```

Comment (denis-tingaikin) [reply pkg/registry/utils/count/nse_client.go:33]: This can be private.

File: pkg/registry/utils/count/nse_client.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	totalRegisterCalls, totalUnregisterCalls, totalFindCalls int32
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/nse_client.go:34]:
```suggestion
	totalRegisterCalls, totalUnregisterCalls, totalFindCalls *int32
```

File: pkg/registry/utils/count/nse_client.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+// Unregisters returns Unregister count
+func (c *NSEClient) Unregisters() int {
+	return int(atomic.LoadInt32(&c.totalUnregisterCalls))
+}
+
+// Finds returns Find count
+func (c *NSEClient) Finds() int {
+	return int(atomic.LoadInt32(&c.totalFindCalls))
+}
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/nse_client.go:94]:
```suggestion
```

File: pkg/registry/utils/count/nse_client.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+)
+
+// NSEClient is a client type for counting Register / Unregister / Find
+type NSEClient struct {
+	totalRegisterCalls, totalUnregisterCalls, totalFindCalls int32
+	registers, unregisters, finds                            map[string]int32
+	mu                                                       sync.Mutex
+}
+
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/nse_client.go:38]:
```suggestion
func NewNetworkServiceEndpointClinet(opts...options) registry.NetworkServiceEndpointClient {
var result = &countNSEClient{
totalRegisterCalls: new(int32)
//TODO totalUnregisterCalls, totalFindCalls
}
for _, opt := range options {
opt(result)
}
return result
}
```
where options
```go
WithRegisters(r *int32) func(*options)
....
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/registry/utils/count/nse_client.go (hunk 7 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go:94]: How do we assert `Register`/`Unregister`/`Find` call count in tests?

Comment (denis-tingaikin) [reply pkg/registry/utils/count/nse_client.go:94]: @Sol-0 Does https://github.com/networkservicemesh/sdk/pull/1202#discussion_r772551028 resolve your question?

**Comment by denis-tingaikin** [comment]
@Sol-0 Please ping me when you've resolved all comments.

File: pkg/registry/common/retry/options.go (hunk 1 above)
Comment (Sol-0) [reply pkg/registry/common/retry/options.go]: Done

File: pkg/registry/common/retry/options.go (hunk 2 above)
Comment (Sol-0) [reply pkg/registry/common/retry/options.go]: Done

File: pkg/registry/common/retry/options.go (hunk 3 above)
Comment (Sol-0) [reply pkg/registry/common/retry/options.go]: Done

File: pkg/registry/utils/count/nse_client.go (hunk 4 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go]: Done

File: pkg/registry/utils/count/nse_client.go (hunk 5 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go]: Done

File: pkg/registry/utils/count/nse_client.go (hunk 6 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go]: Done

File: pkg/registry/utils/count/nse_client.go (hunk 7 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go]: @denis-tingaikin yes, thanks! Done

File: pkg/registry/utils/count/nse_client.go (hunk 8 above)
Comment (Sol-0) [reply pkg/registry/utils/count/nse_client.go:35]: Done

**Comment by Sol-0** [comment]
@denis-tingaikin, comments have been addressed - please, take a look.

File: pkg/registry/utils/count/options.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+		o.totalUnregisterCalls = unregisterCount
+	}
+}
+
+// WithTotalFindCalls sets pointer to number of Find calls
+func WithTotalFindCalls(findCalls *int32) func(*options) {
+	return func(o *options) {
+		o.totalFindCalls = findCalls
+	}
+}
```
Comment (denis-tingaikin) [inline pkg/registry/utils/count/options.go:45]:
```suggestion
type Counter struct {
}
// Registers returns Register call count
func (c *Counter) Registers() int {
	return int(atomic.LoadInt32(&c.totalRegisterCalls))
}

// Unregisters returns Unregister count
func (c *Counter) Unregisters() int {
	return int(atomic.LoadInt32(&c.totalUnregisterCalls))
}

// Finds returns Find count
func (c *Counter) Finds() int {
	return int(atomic.LoadInt32(&c.totalFindCalls))
}
// TODO: add this to the constructor of count servers
```

Comment (Sol-0) [reply pkg/registry/utils/count/options.go]: Done

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**